### ISSUE TITLE
UTF-8 / char / u8 conversion problems

### ISSUE THREAD

**Issue by jcsahnwaldt** [bmacnaughton/u8-swar#1]
A Rust `str` contains UTF-8 bytes, not arbitrary bytes.

This means that `make_u8()` may reject a `str` with fewer than 3 chars if any of the chars are non-ASCII (and thus multi-byte in UTF-8), because `s.len()` returns the length in bytes, not in chars.

It also means that the tests don't work as expected: They try to test `u8` values up to `0xff`, but when `format!()` adds e.g. the `char` value `0xff` to a `str`, the resulting `str` actually contains the two bytes `0xc3` and `0xbf`.

I'm not sure how to fix this. Maybe `make_u8()` should take a `&[u8]` instead of a `&str`?

**Comment by bmacnaughton** [comment]
I don't see a problem here, though I didn't think of the issue that a single unicode char of 3 bytes would pass the length check.

But all the code cares about is that 1) there are from 1 to 3 bytes and 2) that all those bytes are in the ASCII character range from '0' to '9'. I think the test for valid numbers below will detect unicode characters as invalid (though I should add a test for that).

> It also means that the tests don't work as expected: They try to test u8 values up to 0xff, but when format!() adds e.g. the char value 0xff to a str, the resulting str actually contains the two bytes 0xc3 and 0xbf.

I don't understand this point - format is only converting numbers to strings. So 0xFF appears as 255 in the string.

**Comment by jcsahnwaldt** [comment]
I added a few lines to print the char / byte / int values: https://github.com/jcsahnwaldt/u8-swar/blob/main/src/main.rs

Running the tests with `cargo test -- --nocapture  --test-threads 1` should clarify what's going on. I don't think the tests do what you expected them to do... :-)

**Comment by jcsahnwaldt** [comment]
> format is only converting numbers to strings. So 0xFF appears as 255 in the string.

I'm not sure what you mean, but there's no byte with the value `0xFF` / `255` in the string. The char value `0xFF` gets converted to two UTF-8 bytes `0xC3` and `0xBF` which represent the Unicode code point U+00FF, i.e. `ÿ`.
https://play.rust-lang.org/?gist=b93efb054eaaf6c4d3b33d0e9d227429

**Comment by bmacnaughton** [comment]
Yes, you're right. The tests for values greater than 127 are wrong.

Thanks. I'm on my phone right now so can't play with it, but your last
message made it clear.

I just can't test values that aren't a single byte (and don't need to
because they're not digits).

Nice catch!

from my Pixel 8

**Comment by bmacnaughton** [comment]
would  you care to review my fix? https://github.com/bmacnaughton/u8-swar/pull/2

For closure:

>I'm not sure what you mean, but there's no byte with the value

I was looking in the wrong place - there is a test that verified all strings from '0' to '255' were valid. I hadn't associated your comment the later tests that were not testing the byte values I intended.

The last test I added does test with arbitrary values and showed the problem. It's fixed now. It's an interesting problem thinking about processing multiple values in single operations. It's also the basis of bit-counting algorithms.
